sqlx-core/src/postgres/types/ranges/pg_ranges.rs

```diff
@@ -0,0 +1,84 @@
+use crate::{
+    decode::Decode,
+    encode::{Encode, IsNull},
+    postgres::{
+        types::ranges::pg_range::PgRange, PgArgumentBuffer, PgTypeInfo, PgValueRef, Postgres,
+    },
+    types::Type,
+};
+
+macro_rules! impl_pg_range {
+    ($range_name:ident, $type_info:expr, $type_info_array:expr, $range_type:ty) => {
+        #[derive(Clone, Debug, Hash, PartialEq, Eq)]
+        #[repr(transparent)]
+        pub struct $range_name(PgRange<$range_type>);
+
+        impl<'a> Decode<'a, Postgres> for $range_name {
+            fn accepts(ty: &PgTypeInfo) -> bool {
+                <PgRange<$range_type> as Decode<'_, Postgres>>::accepts(ty)
+            }
+
+            fn decode(value: PgValueRef<'a>) -> Result<$range_name, crate::error::BoxDynError> {
+                Ok(Self(Decode::<Postgres>::decode(value)?))
+            }
+        }
+
+        impl<'a> Encode<'a, Postgres> for $range_name {
+            fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> IsNull {
+                <PgRange<$range_type> as Encode<'_, Postgres>>::encode(self.0.clone(), buf)
+            }
+        }
+
+        impl Type<Postgres> for $range_name {
+            fn type_info() -> PgTypeInfo {
+                $type_info
+            }
+        }
+
+        impl Type<Postgres> for [$range_name] {
+            fn type_info() -> PgTypeInfo {
+                $type_info_array
+            }
+        }
+
+        impl Type<Postgres> for Vec<$range_name> {
+            fn type_info() -> PgTypeInfo {
+                $type_info_array
+            }
+        }
+    };
+}
+
+impl_pg_range!(
+    Int4Range,
+    PgTypeInfo::INT4_RANGE,
+    PgTypeInfo::INT4_RANGE_ARRAY,
+    i32
+);
```

Comment:
Now that `Encode::produces` returns `Option<PgTypeInfo>`, it doesn't make sense to implement `Type` for `PgRange<i32>`/`PgRange<i64>`/etc because its `Encode::produces` implementation always returns `None`, which is a contradiction and might lead to confusion. Instead, these manual wrappers were implemented individually

Reply:
My intent was that `produces` is an override. It's not a contradiction.

Reply:
I am sorry, it wasn't my intention. Let me remove this statement.

Reply:
Unfortunately, it isn't possible to override Encode for PgRange due to conflicting implementations. I guess these wrappers make sense.